File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/agspec/ENIntervalNumberTest.java
package uk.nhs.nhsx.diagnosiskeydist.agspec;

import org.junit.Test;

import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import static org.junit.Assert.*;

public class ENIntervalNumberTest {

    @Test
    public void testValidForDateEarlierThanKeyGeneration() {
        Date keyGeneratedAt = utcDate(2020, 7, 7, 0, 0, 0, 0);
        ENIntervalNumber enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(keyGeneratedAt);

        Date beforeKeyGeneration = utcDate(2020, 7, 6, 0,0,0, 0);
        assertFalse(enIntervalNumber.validUntil(beforeKeyGeneration));
    }

    @Test
    public void testValidUntil() {
        Date keyGeneratedAt = utcDate(2020, 7, 7, 0, 0, 0, 0);
        ENIntervalNumber enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(keyGeneratedAt);

        Date threeDaysAfterKeyGeneration = utcDate(2020, 7, 10, 0,0,0, 0);
        Date twentyDaysAfterGeneration = utcDate(2020, 7, 27, 0,0,0, 0);

        assertTrue(enIntervalNumber.validUntil(threeDaysAfterKeyGeneration));
        assertFalse(enIntervalNumber.validUntil(twentyDaysAfterGeneration));
    }

    @Test
    public void testValidUntil2() {
        Date dateNow = ENIntervalNumber.enIntervalNumberFromTimestamp(new Date()).toTimestamp();
        ENIntervalNumber enIntervalNumberNow = ENIntervalNumber.enIntervalNumberFromTimestamp(dateNow);

        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        cal.setTime(dateNow);
        cal.add(Calendar.DATE, 15);
        Date nowPlus15Days = cal.getTime();

        cal.add(Calendar.HOUR_OF_DAY, -2);
        Date nowPlus14DaysMinusTwoHours = cal.getTime();

        assertFalse(enIntervalNumberNow.validUntil(nowPlus15Days));
        assertTrue(enIntervalNumberNow.validUntil(nowPlus14DaysMinusTwoHours));
    }

    @Test
    public void testEnIntervalNumberFromTimestamp() {
        assertEquals(2656614, ENIntervalNumber.enIntervalNumberFromTimestamp(utcDate(2020, 7, 5, 17, 0, 0, 0)).getEnIntervalNumber());
    }

    @Test
    public void testEnIntervalNumberFromTimestampInMillis() {
        assertEquals(2656614, ENIntervalNumber.enIntervalNumberFromTimestampInMillis(utcDate(2020, 7, 5, 17, 0, 0, 0).getTime()).getEnIntervalNumber());
    }

    @Test
    public void testEnIntervalNumberFromTimestampInUnixEpochTime() {
        assertEquals(2656614, ENIntervalNumber.enIntervalNumberFromTimestampInUnixEpochTime(utcDate(2020, 7, 5, 17, 0, 0, 0).getTime() / 1000).getEnIntervalNumber());
    }

    @Test
    public void testToTimestampInUnixEpochTime() {
        assertEquals(1593968400l, ENIntervalNumber.enIntervalNumberFromTimestamp(utcDate(2020, 7, 5, 17, 0, 0, 0)).toTimestampInUnixEpochTime());
    }

    @Test
    public void testToTimestampInMillis() {
        assertEquals(1593968400000l, ENIntervalNumber.enIntervalNumberFromTimestamp(utcDate(2020, 7, 5, 17, 0, 0, 0)).toTimestampInMillis());
    }

    @Test
    public void testToTimestamp() {
        assertEquals(1593968400000l, ENIntervalNumber.enIntervalNumberFromTimestamp(utcDate(2020, 7, 5, 17, 0, 0, 0)).toTimestamp().getTime());
    }

    @Test
    public void testToString() {
        Date keyGeneratedDate = utcDate(2020, 7, 5, 17, 0, 0, 0);
        ENIntervalNumber enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(keyGeneratedDate);

        assertEquals("ENIntervalNumber(2656614: 2020-07-05 17:00)", enIntervalNumber.toString());
    }

    private Date utcDate(int year, int month, int day, int hour, int minute, int second, int millis) {
        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        cal.set(Calendar.YEAR, year);
        cal.set(Calendar.MONTH, month - 1);
        cal.set(Calendar.DAY_OF_MONTH, day);
        cal.set(Calendar.HOUR_OF_DAY, hour);
        cal.set(Calendar.MINUTE, minute);
        cal.set(Calendar.SECOND, second);
        cal.set(Calendar.MILLISECOND, millis);

        return cal.getTime();
    }
}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/agspec/ENIntervalNumberTest.kt
@file:Suppress("UsePropertyAccessSyntax")

package uk.nhs.nhsx.diagnosiskeydist.agspec

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import java.time.Instant

class ENIntervalNumberTest {

    @ParameterizedTest
    @CsvSource(
        "2020-07-07T00:00:00.000Z,2020-07-06T00:00:00.000Z,false",
        "2020-07-07T00:00:00.000Z,2020-07-27T00:00:00.000Z,false",
        "2020-07-07T00:00:00.000Z,2020-07-22T00:00:00.000Z,false",
        "2020-07-07T00:00:00.000Z,2020-07-10T00:00:00.000Z,true",
        "2020-07-07T00:00:00.000Z,2020-07-21T22:00:00.000Z,true"
    )
    fun `valid until`(now: String, validUntil: String, expected: Boolean) {
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(now.asInstant())
        assertThat(enIntervalNumber.validUntil(validUntil.asInstant())).isEqualTo(expected)
    }

    @Test
    fun `create enIntervalNumber from timestamp`() {
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp("2020-07-05T17:00:00.000Z".asInstant())
        assertThat(enIntervalNumber.enIntervalNumber).isEqualTo(2656614L)
    }

    @Test
    fun `create enIntervalNumber from timestamp millis`() {
        val now = "2020-07-05T17:00:00.000Z".asInstant()
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestampInMillis(now.toEpochMilli())
        assertThat(enIntervalNumber.enIntervalNumber).isEqualTo(2656614L)
    }

    @Test
    fun `create enIntervalNumber from timestamp unix epoch time`() {
        val now = "2020-07-05T17:00:00.000Z".asInstant()
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestampInUnixEpochTime(now.epochSecond)
        assertThat(enIntervalNumber.enIntervalNumber).isEqualTo(2656614L)
    }

    @Test
    fun `as TimestampInUnixEpochTime`() {
        val now = "2020-07-05T17:00:00.000Z".asInstant()
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(now)
        assertThat(enIntervalNumber.toTimestampInUnixEpochTime()).isEqualTo(1593968400L)
    }

    @Test
    fun `as TimestampInMillis`() {
        val now = "2020-07-05T17:00:00.000Z".asInstant()
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(now)
        assertThat(enIntervalNumber.toTimestampInMillis()).isEqualTo(1593968400000L)
    }

    @Test
    fun `as Timestamp`() {
        val now = "2020-07-05T17:00:00.000Z".asInstant()
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(now)
        assertThat(enIntervalNumber.toTimestamp().toEpochMilli()).isEqualTo(1593968400000L)
    }

    @Test
    fun `as String`() {
        val now = "2020-07-05T17:00:00.000Z".asInstant()
        val enIntervalNumber = ENIntervalNumber.enIntervalNumberFromTimestamp(now)
        assertThat(enIntervalNumber.toString()).isEqualTo("ENIntervalNumber(2656614: 2020-07-05 17:00)")
    }

    private fun String.asInstant() = Instant.parse(this)
}
